chore: replace yarn with npm

boneskull opened the pull request:
This removes yarn in lieu of npm.

I'm unsure if this affects `@lavamoat/yarn-plugin-allow-scripts` or not.

Two further PRs will build on this.  It is part of the "new release process" effort.

boneskull committed:
chore: reconfigure lerna for releases

This is an experimental change.

It modifies the Lerna configuration to publish independent packages (without `nx`) using conventional commits.  This will create/update CHANGELOGs and create GitHub Releases for each workspace upon publish.

This is what [Appium](https://github.com/appium/appium) does, and I believe the goals for our release process are nearly identical.  Please look at the releases/tags in that repo for what you get.

To publish, run `npm run publish`.

boneskull committed:
chore: convert lockfiles

Used [synp](https://npm.im/synp) to convert from `yarn.lock` to `package-lock.json`.  All resulting `package-lock.json` files are v1 _except_ the root which is v2.

boneskull committed:
chore(ci): update CI to use npm

boneskull committed:
chore: fix lint config

boneskull committed:
chore: attempt to fix lavamoat tests

- fix `test:prep` execution
- using `npm`, node/test/projects/1/node_modules will be symlinked and not under vcs
- pin keccak version
- cleanup `allow-scripts` tests (a bit of DRY, fix misuse of `path.join`, make fs cleanup more consistent)
- fix a problem in the `allow-scripts` tests where _not enough_ cleanup was happening

boneskull committed:
chore: add release-please configuration

This encodes the proper `lerna publish` flags within `lerna.json`.  `release-please` will handle creating the GitHub release and updating the changelog, and we just want Lerna to publish what we have.

These flags are essentially the same thing we have currently, sans the `--yes` flag.

boneskull committed:
chore: replace yarn with npm

This removes all yarn config, caches, and lockfiles, and changes all scripts to use `npm run` instead of yarn`.  It means `lerna bootstrap` is no longer needed, and `lerna` is not needed (but can still be used) to run scripts across workspaces.  There should only be _one_ lockfile: the root `package-lock.json`.

Note that this change impacts what would be done in #625.  Test fixtures, when installed, should no longer create lockfiles.

A note on "resolutions": these become `overrides` and everything can be done in the monorepo root.

This changeset does not modify CI.

boneskull committed:
chore: upgrade all package-lock.json files

boneskull committed:
chore: add commit hooks and prepare script

- adds a pre-commit hook that runs `eslint --fix` on any JS file on the stage
- adds a commit message check that verifies the commit message is in conventional commits format
- adds `husky install` to workspace root `postinstall` script, to be run via `allow-scripts`
- adds a `prepare` lifecycle script which runs the `rebuild` script
- remove `preversion` script

boneskull committed:
chore(ci): add commitlint workflow

naugtur commented on packages/browserify/test/fixtures/secureBundling/lavamoat/node/policy.json:
```diff
@@ -85,6 +85,7 @@
       "globals": {
         "__dirname": true,
         "__filename.slice": true,
+        "console.warn": true,
```
console.warn is there, should have been listed before. Anyway, this change is not undermining the test.

naugtur commented on packages/node/test/projects/2/package.json:
```diff
@@ -1,10 +1,10 @@
   },
   "dependencies": {
-    "keccak": "^3.0.0"
+    "keccak": "3.0.0"
```
makes sense :+1:

naugtur commented on packages/perf/package.json:
```diff
@@ -29,9 +30,6 @@
   },
-  "resolutions": {
-    "serve/boxen/type-fest": "^2.13.0"
-  },
```
@kumavis Do you have any memory of the reason behind type-fest being resolved to 2.13 in perf and to 0.21 pretty much everywhere else?

boneskull replied:
unless something is using `tsc` to build or check types when running things in `perf`, I'm unsure how this could be a problem.  `type-fest` contains no code and does nothing

naugtur reviewed (changes requested):
Now that these are merged I see what I missed before -  reviewing each changeset separately I assumed one of the other ones would contain an update to the RELEASE.md file with the updated process description step-by-step so that everyone knows how to proceed with releasing.

boneskull committed:
docs(release): update release documentation

boneskull committed:
chore(perf): add dummy test script

boneskull committed:
chore: update release-please configuration

This disables the `separate-pull-requests` flag, as enabling it adds too much opportunity for human error to befoul the process.

It also enables the `node-workspace` plugin:

> The node-workspace plugin builds a graph of local node packages configured in release-please-config.json and the dependency relationships between them. It looks at what packages were updated by release-please and updates their reference in other packages' dependencies lists. Even when a particular package was not updated by release-please, if a dependency did have an update, it will be patch bump the package, create a changelog entry, and add it to the list of PR changes. Under the hood, this plugin adapts specific dependency graph building and updating functionality from the popular lerna tool.

Finally, the `prepare` script is removed and a `prepublishOnly` script then rebuilds any artifacts and runs a full test suite before the final publish is green-lit.

boneskull committed:
chore: enable NODE_ENV=production for prepublishOnly
Use cross-env for cross-platform compat. This does not mean other scripts work across platforms, but we should move in that direction.

boneskull committed:
chore(yarn-plugin-allow-scripts): add stub test script

naugtur reviewed (approved):
Thumbsup on the release description. Approving the pr.

(could) I'd prefer to keep the dependency picture (mermaid graph) somewhere and this file seems like a good place for it. 

(later) A checklist for the person doing the release (what to look for in the PR etc) should go into this file too.
